torch/nn/utils/parametrizations.py

```diff
@@ -285,6 +285,84 @@ def orthogonal(module: Module,
     return module
 
 
+class _WeightNorm(Module):
+    def __init__(
+        self,
+        dim: int = 0,
+    ) -> None:
+        super().__init__()
+        if dim is None:
+            dim = -1
+        self.dim = dim
+
+    def forward(self, weight_g, weight_v):
+        return torch._weight_norm(weight_v, weight_g, self.dim)
+
+    def right_inverse(self, weight):
+        # TODO: is the .data necessary?
+        weight_g = torch.norm_except_dim(weight, 2, self.dim).data
+        weight_v = weight.data
```

Comment:
Pretty sure it's not necessary in the first one. Not sure about the second one as it's written.

That being said, `weight_v` shoudl be `weight` divided by `weight_g`. When that's written that way, then I'm pretty sure you don't need the `.data`.

Reply:
Hmm, but that is not what the original code does?

```
        module.register_parameter(name + '_g', Parameter(norm_except_dim(weight, 2, dim).data))
        module.register_parameter(name + '_v', Parameter(weight.data))
        setattr(module, name, fn.compute_weight(module))
```

Reply:
I reread the paper and not dividing is correct. What is confusing in the docs is we refer to a direction vector, but we do not actually represent the direction vector directly (because it is required to orthogonal) so classic weight norm also reparametrizes the direction vector into v (the original weight vector) with the direction vector now represented as `v / || v ||`

Reply:
Right, had a look at the paper and indeed it's as you mention.